x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java

```diff
@@ -5,54 +5,225 @@
+            }
+        }, e -> {
+            ActionListener<EqlSearchResponse> acquiredListener = exclusiveListener.getAndSet(null);
+            if (acquiredListener != null) {
+                // We finished before timeout
+                timeoutHandler.cancel();
+                if (keepOnCompletion) {
+                    storeResults(searchTask,
+                        new EqlStoredResponse(e, threadPool.absoluteTimeInMillis() + keepAlive.getMillis()),
+                        ActionListener.wrap(() -> acquiredListener.onFailure(e)));
+                } else {
+                    taskManager.unregister(searchTask);
+                    acquiredListener.onFailure(e);
+                }
+            } else {
+                // We finished after timeout - saving exception
+                storeResults(searchTask, new EqlStoredResponse(e, threadPool.absoluteTimeInMillis() + keepAlive.getMillis()));
+            }
+        });
+    }
+
+    private void storeResults(EqlSearchTask searchTask, EqlStoredResponse storedResponse) {
+        storeResults(searchTask, storedResponse, ActionListener.wrap(r -> {
+        }, e -> {
+        }));
+    }
+
+    private void storeResults(EqlSearchTask searchTask, EqlStoredResponse storedResponse, ActionListener<Void> finalListener) {
+        try {
+            asyncTaskIndexService.createResponse(searchTask.getAsyncExecutionId().getDocId(),
+                threadPool.getThreadContext().getHeaders(), storedResponse, ActionListener.wrap(
+                    // We should only unregister after the result is saved
+                    resp -> {
+                        taskManager.unregister(searchTask);
+                        logger.trace(() -> new ParameterizedMessage("stored eql search results for [{}]",
+                            searchTask.getAsyncExecutionId().getEncoded()));
+                        finalListener.onResponse(null);
+                    },
+                    exc -> {
+                        taskManager.unregister(searchTask);
+                        Throwable cause = ExceptionsHelper.unwrapCause(exc);
+                        if (cause instanceof DocumentMissingException == false &&
+                            cause instanceof VersionConflictEngineException == false) {
+                            logger.error(() -> new ParameterizedMessage("failed to store eql search results for [{}]",
+                                searchTask.getAsyncExecutionId().getEncoded()), exc);
+                        }
+                        finalListener.onFailure(exc);
+                    }));
+        } catch (Exception exc) {
+            taskManager.unregister(searchTask);
+            logger.error(() -> new ParameterizedMessage("failed to store eql search results for [{}]",
+                searchTask.getAsyncExecutionId().getEncoded()), exc);
+
+        }
```

Comment:
I imagine these utility methods would be useful for every plugin doing async work, as most of the code is around the taskIndexService and taskManager hence why I see value in externalizing them.
It could be in a follow-up PR.
It would be nice to just reuse the boilerplate code (similar to say `ActionListener.wrap`) and only pass in the types either as lambdas or wrapped objects:

`storeResults(asyncTaskIndexService, searchTask, finalListener, task -> searchTask.getAsyncExecutionId().getEncoded())` 

This could sit next to `AsyncExecutionId` in the `async` module so `EqlSearchTask` would just import it and the rest of utilities (like `storeResults`) would just require suppliers for it; then the method above would be even smaller:

`storeResults(asyncTaskIndexService, searchTask, finalListener, task::getAsyncExecutionId)`